Enable convenient unit testing using googletest on POSIX

MaEtUgR opened the pull request:
**Describe problem solved by the proposed pull request**
Part of the description I wrote in #11308

The idea is to enable unit testing for any C++ class of the project with very low overhead. For example if you want to test the class `AttitudeControl` you:

* Create a file `AttitudeControlTest.cpp` next to the classes source files
* Add `px4_add_gtest(SRC AttitudeControlTest.cpp LINKLIBS AttitudeControl)` to the folder's `CMakeLists.txt`
* Start writing your test e.g. :

```
#include <gtest/gtest.h>
#include <AttitudeControl.hpp>

TEST(AttitudeControlTest, AllZeroCase) {
AttitudeControl attitude_control;
matrix::Vector3f rate_setpoint = attitude_control.update(matrix::Quatf(), matrix::Quatf(), 0.f);
EXPECT_EQ(rate_setpoint(0), 0.f);
}
```

When you run `make unit_test` you get the following output (runs all the unit tests):
![first!UNITO-UNDERSCORE!test](https://user-images.githubusercontent.com/4668506/53685343-8799ef80-3d19-11e9-9d8b-ac4feb508af6.PNG)
And it neither needs to rebuild all the libraries you want to test if you're using `px4_sitl_default` anyways nor does gtest need to be loaded when you're not unit testing with it thanks to https://github.com/PX4/Firmware/pull/11533

**Describe your preferred solution**
A clear and concise description of what you have implemented.

**Additional context**

* https://github.com/google/googletest/pull/2101 was necessary such that the global gtest configuration builds under Cygwin.
* A future goal is to be able to run all those test also on the targets.

**Test data / coverage**
Developed and tested on Cygwin Windows toolchain, I will also test on Ubuntu.

dagar commented on cmake/gtest/gtest.cmake:
```diff
@@ -0,0 +1,43 @@
+############################################################################
+
+# Download and unpack googletest at configure time
+configure_file(${CMAKE_CURRENT_LIST_DIR}/CMakeLists.txt.in googletest-download/CMakeLists.txt)
```
I think I'm missing why this has to be done at configure time like this.

MaEtUgR replied:
Because otherwise you don't have the CMakeLists-tree of gtest at configure time and can't configure the dependencies correctly.
I see two options around it:
- You have the source anyways because you add it as a submodule but since all submodules have to be present at anytime I prefered the download only if it's actually needed.
- You hack around and achieve that cmake only needs to configure the dependencies to gtest on build time but I didn't achieve that because you need to wrap it and make assumptions about the structure of gtest. It's easy to screw up the include directories and other things this way.

We can still revisit, this is the first state I find acceptable.

MaEtUgR replied:
btw the method for the download at configure time is the one from the docs, see last point and following description here https://github.com/google/googletest/blob/master/googletest/README.md#incorporating-into-an-existing-cmake-project

dagar commented on cmake/gtest/CMakeLists.txt.in:
```diff
@@ -0,0 +1,18 @@
+
+include(ExternalProject)
+ExternalProject_Add(googletest
+	URL https://github.com/google/googletest/archive/master.zip
```
This should be versioned.

dagar replied:
What about `GIT_REPOSITORY` and `GIT_TAG` instead of `URL`? Newer cmake can do this with a shallow clone.

mcsauder replied:
@MaEtUgR , in case it is helpful, latest is 1.8.1:
```
URL https://github.com/google/googletest/archive/release-1.8.1.tar.gz
URL_HASH SHA1=152b849610d91a9dfa1401293f43230c2e0c33f8
```

MaEtUgR replied:
I totally agree, 
- Cloning the repo works as well but I had the impression it wasn't shallow because it took so much longer compared to just using the zip, would have to check in more detail.
- I didn't use a release for my testing because my pr https://github.com/google/googletest/pull/2101 was just merged and I wanted it in for my Cygwin testing. I'll put a release or at least fixed commit before removing the draft 👍

MaEtUgR replied:
I found out it's possible to download a zip from a specific version without tag from GitHub. I'll add this now.

mcsauder replied:
Indeed ^ , just add this instead of the line you had:
```
URL https://github.com/google/googletest/archive/release-1.8.1.tar.gz
URL_HASH SHA1=152b849610d91a9dfa1401293f43230c2e0c33f8
```

MaEtUgR replied:
I chose `URL https://github.com/google/googletest/archive/8b6d3f9c4a774bef3081195d422993323b6bb2e0.zip`. It's also a version from the master branch just without a tag.

dagar commented on CMakeLists.txt:
```diff
@@ -413,6 +413,20 @@ if (NOT EXTERNAL_MODULES_LOCATION STREQUAL "")
+# Testing - Automatic unit and integration testing with CTest
+#
+
+option(unit_testing "Configure unit test targets" OFF)
```
Can the existing cmake testing facilities be used directly?

`enable_testing()` - https://cmake.org/cmake/help/v3.8/command/enable_testing.html?highlight=i
Then you can check `BUILD_TESTING `.https://github.com/PX4/Firmware/blob/d42bb01e0ca83813f1d8a96576c6fb9244bb7a19/platforms/posix/CMakeLists.txt#L117

MaEtUgR replied:
I wanted to do that but they are used already for the SITL integration tests. I'd like to be able to run the tests separately since the SITL tests take so much longer. I planned to have a structured test name and use `ctest -R` to filter the correct ones. So you think distinguishing build time doesn't make sense? I could agree with that. Then gtest is always needed for tests in general and it all depends on `BUILD_TESTING`.

MaEtUgR replied:
Ah and testing is always enabled for sitl... https://github.com/PX4/Firmware/blob/fc7c7ac20630951b17e8644ccc175b9c4e743409/CMakeLists.txt#L303-L305
Can I make it optional?

dagar replied:
> since the SITL tests take so much longer

Are you talking about the SITL unit tests? Each is roughly 1-3 seconds.
Example: http://ci.px4.io:8080/blue/organizations/jenkins/PX4%2FFirmware/detail/master/722/pipeline/21

![Screen Shot 2019-03-09 at 3 45 14 PM](https://user-images.githubusercontent.com/84712/54077129-62116680-4282-11e9-9e26-6bf13952a92b.png)

MaEtUgR replied:
Yes, I expected the basic tests to run two order of magnitudes faster. Is that impatient? 😇 I think it's just starting the simulator binary that takes the big part of that time, otherwise the smallest tests would not all have around the same duration.
Anyways I think we should allow to run the unit tests separately since I presume they'll finish a lot faster until we write a crazy amount of tests or huge loops.

MaEtUgR replied:
I'll just add all tests and the gtest library fetching with `enable_testing()` and hence `BUILD_TESTING` but make enabling tests optional for posix. Then I can still filter the tests if I only want to run the quick unit tests e.g. with `make tests unit`.

MaEtUgR commented on src/include/visibility.h:
```diff
@@ -86,7 +86,7 @@
 #include <cstdlib>
 #endif
-#pragma GCC poison exit
+//#pragma GCC poison exit
```
I also have a question: I didn't figure out what to do about exit being poisoned but gtest using exit. Should I reenable it, remove visibility.h from the unit test builds? Is there an easy answer?

dagar replied:
This gets into how the binary is built. If we can get away with building separate test binaries or if the test itself needs to be an "app" within PX4.

My initial thought is that the options are
1.  build should be structured such that the unit test executable is not inheriting these flags at all 
2. each test is a module in PX4 and run like the existing unit tests. https://github.com/PX4/Firmware/blob/master/posix-configs/SITL/init/test/test_template.in#L24

MaEtUgR replied:
I would opt for 1. now and make sure to look into the possibilities of 2. as well at some point to be able to run the tests on the microcontroller targets.

MaEtUgR replied:
I figured that one one by removing the global `-include visibility.h` compile option in the gtest cmake. This flag should not be kept global in the future but rather added to the targets where it's necessary as discussed below.

dagar commented on cmake/gtest/gtest.cmake:
```diff
@@ -41,3 +41,9 @@ endif()
 add_subdirectory(${CMAKE_CURRENT_BINARY_DIR}/googletest-src ${CMAKE_CURRENT_BINARY_DIR}/googletest-build EXCLUDE_FROM_ALL)
+
+# Remove visibility.h from the compile flags for gtest because of poisoned exit()
+get_target_property(GTEST_COMPILE_FLAGS gtest COMPILE_OPTIONS)
```
Longer term what we should do system wide is only apply our default flags, defines, includes, etc on a per target basis. Then in the gtest case we won't have this problem in the first place.

MaEtUgR replied:
That's exactly what I thought but I didn't want to dig over the whole build setup right now. We can still do this. Targets are definitely the way to go in contrast to including everything everywhere 😛. While blaming I saw that you are constantly cleaning up parts of the cmake configuration step by step. 👍

MaEtUgR commented:
rebased after merging #11308

MaEtUgR commented:
Rebased and started to combine the unit test framework with camkes `BUILD_TESTING` after the comment here: https://github.com/PX4/Firmware/pull/11573#discussion_r261839107 Next up is filtering for the different tests such that it's possible to specify what test you want to run for local test-driven development and to exclude WIP tests in CI.

mcsauder commented:
@MaEtUgR , Its not clear to me how to build/run the tests with your latest commit, could you clue me in? :)

Your previous [commit](https://github.com/PX4/Firmware/pull/11573/commits/dec516f31f237e36da8b934cea6f4a21e818f4c2) and others before it run great on Ubuntu 18.10:

![image](https://user-images.githubusercontent.com/2497951/56147111-da93d300-5f64-11e9-880d-0e2f01f9f667.png)

MaEtUgR commented:
@mcsauder Sorry the latest commit at the time you wrote your comment was an unfinished temporary dump of ongoing tests that's why the commit message started with `TEMP:`. It ran certain tests double when commanding `make tests`. I replaced it over the weekend now and added the possibility to filter tests with `make tests TESTFILTER=unit`. I still need to clean up the `unit_test` make target because it will get obsolete with one test target and filtering.

MaEtUgR commented:
I rebased again. Since I would now start writing useful unit tests I would say this pr is ready for review. One thing I already noticed is the license text is not in every file, I'll fix that quickly.

MaEtUgR commented:
@dagar Any chance to get the test framework in? I'm willing to continue to work on how it's embedded in cmake but would appreciate if I was able to start to write unit tests. The tests are now included in `make tests` and are hence called in CI.

RomanBapst commented:
This is quite useful!

mcsauder commented:
Nice work @MaEtUgR !

MaEtUgR commented:
@dagar rebased on rc1
